File: pastepad/modelo.py
# -*- coding: utf-8 -*-
"""Los datos y sus reglas. Sin tkinter: todo lo de aqui se puede probar
sin abrir una ventana."""

import json
import os
import time

from . import config as cfg


# ------------------------------------------------------------ archivos

def _leer(ruta, defecto):
    try:
        with open(ruta, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return defecto


def _escribir(ruta, datos):
    """Escribe primero en un archivo aparte y luego lo mueve.

    Si el programa se cierra a mitad de la escritura, el archivo bueno
    sigue intacto en vez de quedar cortado.
    """
    temporal = ruta + ".tmp"
    try:
        with open(temporal, "w", encoding="utf-8") as f:
            json.dump(datos, f, ensure_ascii=False, indent=1)
        os.replace(temporal, ruta)
        return True
    except Exception:
        try:
            os.remove(temporal)
        except Exception:
            pass
        return False


def fragmento(texto, fuente=None, tam=None, negrita=0, cursiva=0,
              subrayado=0, color=None):
    """Un trozo de texto con su formato. Un snippet es una lista de estos."""
    return {"t": texto,
            "f": fuente or cfg.FUENTE_DEF,
            "s": tam or cfg.TAM_DEF,
            "b": negrita, "i": cursiva, "u": subrayado,
            "c": color or cfg.COLOR_DEF}


def texto_de(fragmentos):
    return "".join(f["t"] for f in fragmentos)


def una_linea(texto, tope=52):
    """Resumen de una linea. Corta antes de separar en palabras: con
    textos de miles de lineas, hacerlo al reves cuesta casi un segundo."""
    crudo = texto[:tope * 4]
    limpio = " ".join(crudo.split())
    if len(texto) > len(crudo) or len(limpio) > tope:
        return limpio[:tope] + "..."
    return limpio


# ------------------------------------------------------------ enlaces

def es_enlace(texto):
    """True si el texto es una direccion web y nada mas.

    Solo si el texto entero es el enlace: un parrafo que menciona una
    url de pasada no cuenta, porque abrirlo no seria lo que el usuario
    espera al hacer clic.
    """
    if not texto:
        return False
    t = texto.strip()
    if " " in t or "\n" in t or len(t) > 2000:
        return False
    return t.lower().startswith(("http://", "https://", "www."))


def url_de(texto):
    """La direccion lista para abrir en el navegador."""
    t = texto.strip()
    return "https://" + t if t.lower().startswith("www.") else t


def dominio_de(texto):
    """El dominio suelto, para mostrarlo debajo del titulo."""
    t = url_de(texto)
    for prefijo in ("https://", "http://"):
        if t.startswith(prefijo):
            t = t[len(prefijo):]
            break
    if t.startswith("www."):
        t = t[4:]
    return t.split("/")[0][:60]


# ------------------------------------------------------------ plantillas

def campos_de(texto):
    """Los [[campos]] de una plantilla, en orden y sin repetir."""
    campos, resto = [], texto
    while "[[" in resto and "]]" in resto:
        i = resto.index("[[")
        j = resto.index("]]", i)
        nombre = resto[i + 2:j].strip()
        if nombre and nombre not in campos:
            campos.append(nombre)
        resto = resto[j + 2:]
    return campos


def rellenar(fragmentos, valores):
    salida = []
    for f in fragmentos:
        t = f["t"]
        for clave, valor in valores.items():
            t = t.replace("[[%s]]" % clave, valor)
        copia = dict(f)
        copia["t"] = t
        salida.append(copia)
    return salida


# ------------------------------------------------------------ almacen

class Almacen:
    """Todo el estado de la aplicacion en un solo sitio.

    Antes cada parte de la interfaz leia y escribia los archivos por su
    cuenta; ahora pasan por aqui, que es lo unico que toca el disco.
    """

    def __init__(self):
        self.prefs = _leer(cfg.RUTA_PREFS, {})
        self.hist = _leer(cfg.RUTA_HIST, [])
        datos = _leer(cfg.RUTA_DATOS, {})
        self.carpetas = datos.get("categorias", [])
        self.snippets = datos.get("snippets", [])
        for s in self.snippets:
            if "runs" not in s:
                s["runs"] = [fragmento(s.get("texto", ""))]

    # ---- preferencias

    def pref(self, clave, defecto=None):
        return self.prefs.get(clave, defecto)

    def poner_pref(self, clave, valor):
        self.prefs[clave] = valor
        _escribir(cfg.RUTA_PREFS, self.prefs)

    # ---- guardados

    def guardar_datos(self):
        _escribir(cfg.RUTA_DATOS, {"categorias": self.carpetas,
                                   "snippets": self.snippets})

    def guardar_hist(self):
        _escribir(cfg.RUTA_HIST, self.hist)

    # ---- carpetas

    def crear_carpeta(self, nombre):
        if nombre and nombre not in self.carpetas:
            self.carpetas.append(nombre)
            self.guardar_datos()
            return True
        return False

    def renombrar_carpeta(self, viejo, nuevo):
        if not nuevo or nuevo in self.carpetas or viejo not in self.carpetas:
            return False
        self.carpetas[self.carpetas.index(viejo)] = nuevo
        for s in self.snippets:
            if s["categoria"] == viejo:
                s["categoria"] = nuevo
        self.guardar_datos()
        return True

    def contenido_de(self, carpeta):
        return [s for s in self.snippets if s["categoria"] == carpeta]

    def borrar_carpeta(self, carpeta):
        """Se lleva la carpeta y todo lo que tenga dentro."""
        dentro = self.contenido_de(carpeta)
        for s in dentro:
            self.snippets.remove(s)
        if carpeta in self.carpetas:
            self.carpetas.remove(carpeta)
        self.guardar_datos()
        return len(dentro)

    # ---- snippets

    def anadir_snippet(self, snippet):
        self.crear_carpeta(snippet["categoria"])
        self.snippets.append(snippet)
        self.guardar_datos()

    def reemplazar_snippet(self, viejo, nuevo):
        try:
            self.snippets[self.snippets.index(viejo)] = nuevo
        except ValueError:
            return False
        self.crear_carpeta(nuevo["categoria"])
        self.guardar_datos()
        return True

    # ---- historial

    def anotar(self, entrada):
        """Mete algo copiado al principio, debajo de los fijados."""
        for x in self.hist[:4]:
            if entrada["tipo"] == "texto" and \
                    x.get("texto") == entrada.get("texto"):
                return False
        fijados = sum(1 for x in self.hist if x.get("pin"))
        self.hist.insert(fijados, entrada)
        self._recortar()
        self.guardar_hist()
        return True

    def _recortar(self):
        """Deja solo las ultimas MAX_HIST sueltas. Las fijadas no cuentan."""
        libres = [x for x in self.hist if not x.get("pin")]
        for viejo in libres[cfg.MAX_HIST:]:
            self._borrar_imagen(viejo)
            self.hist.remove(viejo)

    @staticmethod
    def _borrar_imagen(entrada):
        if entrada.get("tipo") == "imagen":
            try:
                os.remove(entrada.get("ruta", ""))
            except Exception:
                pass

    def fijar(self, entrada):
        entrada["pin"] = not entrada.get("pin")
        self.guardar_hist()

    def borrar(self, elemento):
        """Sirve para los dos tipos: el historial lleva 'tipo', los
        guardados no."""
        if elemento.get("tipo"):
            self._borrar_imagen(elemento)
            try:
                self.hist.remove(elemento)
            except ValueError:
                return False
            self.guardar_hist()
        else:
            try:
                self.snippets.remove(elemento)
            except ValueError:
                return False
            self.guardar_datos()
        return True

    def borrar_varios(self, elementos):
        return sum(1 for e in elementos if self.borrar(e))

    def vaciar_historial(self):
        """Los fijados sobreviven: para eso estan."""
        for x in self.hist:
            if not x.get("pin"):
                self._borrar_imagen(x)
        self.hist = [x for x in self.hist if x.get("pin")]
        self.guardar_hist()

    def hist_ordenado(self):
        return ([x for x in self.hist if x.get("pin")] +
                [x for x in self.hist if not x.get("pin")])

    def guardar_imagen(self, datos_dib, a_bmp):
        try:
            os.makedirs(cfg.DIR_IMG, exist_ok=True)
            ruta = os.path.join(cfg.DIR_IMG,
                                "img_%d.bmp" % int(time.time() * 1000))
            with open(ruta, "wb") as f:
                f.write(a_bmp(datos_dib))
            return self.anotar({"tipo": "imagen", "ruta": ruta})
        except Exception:
            return False

File: csharp/Pastepad.Nucleo/Modelo.cs
namespace Pastepad.Nucleo;

/// <summary>
/// Las reglas de los datos, sin nada de interfaz. Todo lo de aqui se
/// puede probar sin abrir una ventana.
/// </summary>
public static class Modelo
{
    public static Fragmento CrearFragmento(
        string texto,
        string? fuente = null,
        int? tam = null,
        int negrita = 0,
        int cursiva = 0,
        int subrayado = 0,
        string? color = null) => new()
        {
            T = texto,
            F = fuente ?? Config.FuenteDef,
            S = tam ?? Config.TamDef,
            B = negrita,
            I = cursiva,
            U = subrayado,
            C = color ?? Config.ColorDef,
        };

    public static string TextoDe(IEnumerable<Fragmento> fragmentos) =>
        string.Concat(fragmentos.Select(f => f.T));

    /// <summary>
    /// La primera linea con algo escrito, con los espacios de dentro
    /// normalizados. Ni corta ni añade: lo que devuelve esta contenido
    /// tal cual en el texto del usuario.
    ///
    /// Es la que vale para GUARDAR. Para la pantalla esta
    /// <see cref="UnaLinea"/>, que marca el recorte con puntos
    /// suspensivos — y esos puntos no pueden acabar en un archivo.
    /// </summary>
    public static string PrimeraLinea(string texto)
    {
        if (string.IsNullOrEmpty(texto)) return "";

        foreach (var linea in texto.Split('\n'))
        {
            string limpio = string.Join(" ", linea.Split(
                (char[]?)null, StringSplitOptions.RemoveEmptyEntries));

            if (limpio.Length > 0) return limpio;
        }

        return "";
    }

    /// <summary>
    /// Un guardado nuevo a partir de lo que el usuario escribio.
    ///
    /// Vive aqui y no en cada dialogo porque el titulo es un dato que va
    /// a snippets.json y tiene que decidirse en un solo sitio. Se
    /// guardaba resumido con <see cref="UnaLinea"/>, y el resumen dejaba
    /// los puntos suspensivos DENTRO del archivo: quien leyera el titulo
    /// se encontraba un texto que el usuario nunca escribio. Acortarlo
    /// para que quepa en la fila es cosa de la interfaz, que es la unica
    /// que sabe cuanto cabe.
    /// </summary>
    public static Snippet CrearSnippet(string texto, string categoria) => new()
    {
        Titulo = PrimeraLinea(texto),
        Categoria = categoria,
        Runs = [CrearFragmento(texto)],
    };

    /// <summary>
    /// Resumen de una linea, para ensenar. Corta antes de separar en
    /// palabras: con textos de miles de lineas, hacerlo al reves cuesta
    /// casi un segundo.
    /// </summary>
    public static string UnaLinea(string texto, int tope = 52)
    {
        if (string.IsNullOrEmpty(texto)) return "";

        int corte = tope * 4;
        string crudo = texto.Length > corte ? texto[..corte] : texto;

        string limpio = string.Join(" ", crudo.Split(
            (char[]?)null, StringSplitOptions.RemoveEmptyEntries));

        if (texto.Length > crudo.Length || limpio.Length > tope)
            return limpio[..Math.Min(tope, limpio.Length)] + "...";

        return limpio;
    }

    // ---------------------------------------------------------- enlaces

    /// <summary>
    /// True si el texto es una direccion web y nada mas. Solo si el
    /// texto entero es el enlace: un parrafo que menciona una url de
    /// pasada no cuenta, porque abrirlo no seria lo que el usuario
    /// espera al hacer clic.
    /// </summary>
    public static bool EsEnlace(string? texto)
    {
        if (string.IsNullOrEmpty(texto)) return false;

        string t = texto.Trim();

        if (t.Contains(' ') || t.Contains('\n') || t.Length > 2000)
            return false;

        string bajo = t.ToLowerInvariant();

        return bajo.StartsWith("http://", StringComparison.Ordinal)
            || bajo.StartsWith("https://", StringComparison.Ordinal)
            || bajo.StartsWith("www.", StringComparison.Ordinal);
    }

    /// <summary>La direccion lista para abrir en el navegador.</summary>
    public static string UrlDe(string texto)
    {
        string t = texto.Trim();

        return t.StartsWith("www.", StringComparison.OrdinalIgnoreCase)
            ? "https://" + t
            : t;
    }

    /// <summary>El dominio suelto, para mostrarlo debajo del titulo.</summary>
    public static string DominioDe(string texto)
    {
        string t = UrlDe(texto);

        foreach (var prefijo in new[] { "https://", "http://" })
        {
            if (t.StartsWith(prefijo, StringComparison.Ordinal))
            {
                t = t[prefijo.Length..];
                break;
            }
        }

        if (t.StartsWith("www.", StringComparison.Ordinal))
            t = t[4..];

        string dominio = t.Split('/')[0];

        return dominio.Length > 60 ? dominio[..60] : dominio;
    }

    // ------------------------------------------------------- plantillas

    /// <summary>
    /// Los [[campos]] de una plantilla, en orden y sin repetir.
    /// </summary>
    public static List<string> CamposDe(string texto)
    {
        var campos = new List<string>();
        string resto = texto;

        while (true)
        {
            int i = resto.IndexOf("[[", StringComparison.Ordinal);
            if (i < 0) break;

            int j = resto.IndexOf("]]", i, StringComparison.Ordinal);

            // La version en Python no comprobaba esto y reventaba con un
            // texto como "]] [[x", donde el cierre va antes que la
            // apertura. Aqui simplemente se deja de buscar.
            if (j < 0) break;

            string nombre = resto[(i + 2)..j].Trim();

            if (nombre.Length > 0 && !campos.Contains(nombre))
                campos.Add(nombre);

            resto = resto[(j + 2)..];
        }

        return campos;
    }

    public static List<Fragmento> Rellenar(
        IEnumerable<Fragmento> fragmentos,
        IReadOnlyDictionary<string, string> valores)
    {
        var salida = new List<Fragmento>();

        foreach (var f in fragmentos)
        {
            string t = f.T;

            foreach (var (clave, valor) in valores)
                t = t.Replace("[[" + clave + "]]", valor, StringComparison.Ordinal);

            salida.Add(new Fragmento
            {
                T = t,
                F = f.F,
                S = f.S,
                B = f.B,
                I = f.I,
                U = f.U,
                C = f.C,
            });
        }

        return salida;
    }
}
